gdsfactory/functions.py

```diff
@@ -70,27 +70,28 @@ def extract(
     Args:
         component: to extract the layers from.
         layers: list of layers to extract.
-        recursive: if True, extracts layers recursively and returns a flattened Component.
+        recursive: if True, extracts the shapes recursively.
     """
     from gdsfactory.pdk import get_layer_tuple
 
-    c = component.dup()
-    if recursive:
-        c.flatten()
+    c = gf.Component()
 
     layer_tuples = [get_layer_tuple(layer) for layer in layers]
-    component_layers = get_layers(c)
+    component_layers = get_layers(component)
 
     for layer_tuple in layer_tuples:
         if layer_tuple not in component_layers:
             warnings.warn(
-                f"Layer {layer_tuple} not found in component {component.name!r} layers."
+                f"Layer {layer_tuple} not found in component {component.name!r} layers. {component_layers}"
```

Comment:
just create a set here?

`for layer_tuple in set(layer_tuples) & set(component_layers)`